In [40]:
# Librerías necesarias para nuestro análisis:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import matplotlib.patches as mpatches
import plotly_express as px
import plotly.graph_objects as go
import os

# Modelos clasificatorios:
from sklearn.metrics import cohen_kappa_score
from sklearn.metrics import f1_score
from sklearn.metrics.pairwise import cosine_distances
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import make_pipeline
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score


COEFICIENTES PARA ESTUDIAR LA CALIDAD DE LOS DATOS DEL DATASET EN TÉRMINOS ANALIZAR Y PREDICIÓN 

In [22]:
# Cargar el dataset a analizar:
df_clean = pd.read_csv('clean_data.csv')

In [23]:
# Probamos el coeficiente Kappa
x = df_clean[['Gender', 'Age', 'EDUC', 'SES', 'MMSE', 'CDR', 'eTIV']] #escogemos todas las columnaa menos la ultima
y = df_clean['Group'] #escogemos solo la ultima

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.25)

modelo = LogisticRegression()
modelo.fit(x_train, y_train)

y_pred = modelo.predict(x_test)

kappa = cohen_kappa_score(y_test, y_pred)

print(f"El coeficiente kappa es: {kappa:.4f}")


El coeficiente kappa es: 0.8444


C:\Users\campo\AppData\Roaming\Python\Python311\site-packages\sklearn\linear_model\_logistic.py:460: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


Conclusión: 0.8444 demuestra un porcentaje muy bueno, por lo que sugiere que los clasificadores están alineados en sus predicciones.
Valores positivos y cercanos a 1 señalan mayor concordancia que la esperada por azar.

In [24]:
# Distancia del coseno:

from sklearn.metrics.pairwise import cosine_distances

vector1 = df_clean.iloc[0, :-1].values
vector2 = df_clean.iloc[1, :-1].values

distancia = cosine_distances([vector1], [vector2])[0, 0]

print(f"Distancia del coseno: {distancia}")

Distancia del coseno: 1.913671443531939e-10


El resultado obtenido por la distancia del coseno está en notación científica, por lo que el valor es bastante pequeño, lo que indica que la distancia entre los vectores es cercana, por lo que hay una alta similitud, asi que la similitud entre los datos puede ser positiva.

In [35]:
#  Calculamos la medida F1


from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import make_pipeline

x = df_clean.iloc[:, :-1]
y = df_clean.iloc[:, -1]

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.25)

# Crear un pipeline que primero escala los datos y luego aplica la Regresión Logística. 
# También aumenta el número máximo de iteraciones permitidas en el modelo de Regresión Logística a 1000.
modelo = make_pipeline(StandardScaler(), LogisticRegression(max_iter=1000))
modelo.fit(x_train, y_train)

y_pred = modelo.predict(x_test)

# Generar classification report
report = classification_report(y_test, y_pred)

print(report)


              precision    recall  f1-score   support

           0       0.86      0.98      0.92        45
           1       0.86      1.00      0.92        36
           2       1.00      0.08      0.14        13

    accuracy                           0.86        94
   macro avg       0.91      0.68      0.66        94
weighted avg       0.88      0.86      0.81        94



Usamos una medida F-1 para evaluar el rendimiento del modelo.  Con una precision del 86%, indiga un porcentaje de predicciones correctas bueno.
La exhaustivvidad del recall es del 98% para la clase 0 y del 100% para la clase 1. Esto indica que el modelo identifica correctamente la mayoría de los casos positivos.

In [41]:
# Vamos a probar con un modelo más avanzado de Árbol de decisión


x = df_clean.iloc[:, :-1]  # Select all columns except the last one as features
y = df_clean.iloc[:, -1]  # Select the last column as the target variable

# Dividir los datos en training set and test set
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.25)

# Crear un clasificador de árbol de decisión
clf = DecisionTreeClassifier()

# Train the classifier with the training data
clf.fit(x_train, y_train)

# Predict the target variable for the test data
y_pred = clf.predict(x_test)

# Calculate the accuracy of the model
accuracy = accuracy_score(y_test, y_pred)

print(f'Precisión: {accuracy}')

Precisión: 0.9148936170212766
